Modularize yang_parse_tree_paths

ffoulkes opened the pull request:
- Refactor `yang_parse_tree_paths.cc` into multiple source files.
- Move all yang files from `hal/lib/common` to `hal/lib/yang`.
- Update Bazel build to support new yang_parse_tree files.

ffoulkes committed:
Modularize yang_parse_tree_paths

- Divided yang_parse_tree_paths.cc into multiple source files,
  to make it easier to maintain.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Move yang_parse_tree source files to hal/lib/yang directory

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Refactor yang_parse_tree_interface

- Moved setup functions to the stratum::hal::yang:interface namespace
  and created a header file to expose them externally.

- Moved YangParseTreePaths::AddSubtreeInterface to its own source file.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes committed:
Add Bazel support for modularized yang_parse_tree files

Signed-off-by: Derek G Foster <derek.foster@intel.com>

pudelkoM commented:
Is there a writeup on the reasoning behind this change? I want to understand the background  first, before commenting on the details. 
Once thing to note though; huge refactorings are _extremely_ difficult to merge upstream, especially once there are other changes mixed in.

ffoulkes committed:
Document yang_parse_tree_paths refactoring

- Added some brief header comments to the new files.
- Added a project specification to the `docs` directory.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes commented:
> Is there a writeup on the reasoning behind this change? I want to understand the background first, before commenting on the details. Once thing to note though; huge refactorings are _extremely_ difficult to merge upstream, especially once there are other changes mixed in.

I have added a project document to the `docs` directory in the PR. I don't know if this kosher, but it seemed in keeping with the spirit of open source to make this available to the community. I can omit it when/if we upstream this change to the mother ship.

ffoulkes committed:
Add DPDK-specific YangParseTreePaths support

- Ported DPDK-specific functions from yang_parse_tree_paths.cc in
  P4-OVS. Stubbed out the functionality to forward the configuration
  values to the SDE, which has not yet been implemented.

- Corrected include guards in the header files that were moved to
  the stratum/hal/lib/yang directory.

Signed-off-by: Derek G Foster <derek.foster@intel.com>

ffoulkes commented:
Added DPDK-specific files to YangParseTreePaths. `dpdk_add_subtree_interface` is substituted for `yang_add_subtree_interface`, and `dpdk_parse_tree_interface` is added to the build, to define the new DPDK OpenConfig parameters. This was the original impetus for modularizing `yang_parse_tree_paths` .

ffoulkes committed:
Update YangParseTreePaths refactoring document

Signed-off-by: Derek G Foster <derek.foster@intel.com>
